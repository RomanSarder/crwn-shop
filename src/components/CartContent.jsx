import React from 'react'
import styled from 'styled-components'

import Button from './Button'

var StyledCartContent = styled.div`
    position: absolute;
    width: 24rem;
    height: 34rem;
    display: flex;
    flex-direction: column;
    padding: 2rem;
    border: 1px solid black;
    background-color: white;
    top: 8rem;
    right: 0;
    z-index: 5;

    .cart-items {
        height: 24rem;
        display: flex;
        flex-direction: column;
        overflow: scroll;
    }

    ${Button} {
        margin-top: auto;
    }
`

export default function CartContent() {
    return (
        <StyledCartContent>
            <div className="cart-items"></div>
            <Button>Checkout</Button>
        </StyledCartContent>
    )
}
