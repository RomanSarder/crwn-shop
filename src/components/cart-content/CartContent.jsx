import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'
import styled from 'styled-components'
import { toggleCart } from '../../store/cart/actions'
import { selectCartItems, selectCartItemsTotalPrice } from '../../store/cart/selectors'

import Button from '../button/Button'
import CartItem from '../CartItem'
import TotalPrice from '../styled/TotalPrice'

var StyledCartContent = styled.div`
    position: absolute;
    display: flex;
    width: 36rem;
    height: 24rem;
    flex-direction: column;
    padding: 2rem;
    border: 1px solid black;
    background-color: white;
    top: 8rem;
    right: 0;
    z-index: 5;

    .cart-items {
        height: 85%;
        display: flex;
        flex-direction: column;
        overflow: scroll;
    }

    .checkout-btn {
        margin-top: auto;
    }

    .empty-message {
        font-size: 2.2rem;
        margin: 5rem auto;
        font-weight: bold;
    }

    .total {
        margin-left: auto;
    }

    @media (min-width: 768px) {
        height: 34rem;
        width: 42rem;
    }

    @media (min-width: 992px) { 
        width: 48rem;
        height: 38rem;
    }
`

function _renderCartItems (items) {
    return items.map(function mapCartItemsToComponents (item) {
        return <CartItem key={item.id} item={item} />
    })
}

export default function CartContentContainer () {
    var cartItems = useSelector(selectCartItems)
    var totalPrice = useSelector(selectCartItemsTotalPrice)
    var history = useHistory()
    var dispatch = useDispatch()

    function checkout () {
        history.push('/checkout')
        dispatch(toggleCart())
    }

    return <CartContent onCheckout={checkout} total={totalPrice} items={cartItems} />
}

export function CartContent({ onCheckout, total, items, renderCartItems = _renderCartItems }) {
    return (
        <StyledCartContent>
            <div className="cart-items">
                { items.length ? 
                    renderCartItems(items) :
                    <span className="empty-message">Your cart is empty</span> }
            </div>
            <TotalPrice className="total">Total: ${total}</TotalPrice>
            <Button onClick={onCheckout} className="checkout-btn">
                Go To Checkout
            </Button>
        </StyledCartContent>
    )
}
