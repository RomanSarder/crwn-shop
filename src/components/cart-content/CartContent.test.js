import React from 'react'
import { render } from '@testing-library/react'
import { CartContent } from './CartContent'
import userEvent from '@testing-library/user-event';

jest.mock('../CartItem', () => () => (<div>Hello World</div>))

const testCartItems = [
    {
        id: 1,
        imageUrl: 'test1',
        name: 'Test Item 1',
        price: 10,
        quantity: 1
    },
    {
        id: 2,
        imageUrl: 'test2',
        name: 'Test Item 2',
        price: 20,
        quantity: 2
    }
]

it('should render proper cart content with mocked cart items', () => {
    var onCheckoutMock = jest.fn()
    const mockTotalValue = 5

    var { container } = render(<CartContent items={testCartItems} onCheckout={onCheckoutMock} total={mockTotalValue}/>)

    expect(container.firstChild).toMatchSnapshot()
})

it('should have checkout button working', () => {
    var onCheckoutMock = jest.fn()
    const mockTotalValue = 5

    var { getByText } = render(<CartContent items={testCartItems} onCheckout={onCheckoutMock} total={mockTotalValue}/>)

    var button = getByText('Go To Checkout')

    userEvent.click(button)

    expect(onCheckoutMock).toHaveBeenCalledTimes(1)
})

it('should render proper number of cart items', () => {
    var onCheckoutMock = jest.fn()
    const mockTotalValue = 5

    var { getAllByText } = render(<CartContent items={testCartItems} onCheckout={onCheckoutMock} total={mockTotalValue}/>)

    expect(getAllByText('Hello World').length).toEqual(2)
})

it('should properly render cart items', () => {
    var onCheckoutMock = jest.fn()
    var renderCartItemsMock = jest.fn()
    const mockTotalValue = 5

    render(<CartContent items={testCartItems} onCheckout={onCheckoutMock} total={mockTotalValue} renderCartItems={renderCartItemsMock}/>)

    expect(renderCartItemsMock).toHaveBeenCalledWith(testCartItems)
})