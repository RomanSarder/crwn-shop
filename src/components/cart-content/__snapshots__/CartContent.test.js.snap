// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should properly render with redux state 1`] = `
.c2 {
  min-width: 16.5rem;
  width: auto;
  height: 5rem;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 5rem;
  padding: 0 3.5rem 0 3.5rem;
  font-size: 1.5rem;
  background-color: black;
  color: white;
  text-transform: uppercase;
  font-family: 'Open Sans Condensed';
  font-weight: bolder;
  cursor: pointer;
  border: none;
}

.c2:hover {
  background-color: white;
  color: black;
  border: 1px solid black;
}

.c1 {
  text-transform: uppercase;
  font-weight: 700;
  font-size: 2.4rem;
}

.c0 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 36rem;
  height: 24rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 2rem;
  border: 1px solid black;
  background-color: white;
  top: 8rem;
  right: 0;
  z-index: 5;
}

.c0 .cart-items {
  height: 85%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: scroll;
}

.c0 .checkout-btn {
  margin-top: auto;
}

.c0 .empty-message {
  font-size: 2.2rem;
  margin: 5rem auto;
  font-weight: bold;
}

.c0 .total {
  margin-left: auto;
}

@media (min-width:768px) {
  .c0 {
    height: 34rem;
    width: 42rem;
  }
}

@media (min-width:992px) {
  .c0 {
    width: 48rem;
    height: 38rem;
  }
}

<div
  class="c0"
>
  <div
    class="cart-items"
  >
    <div>
      Hello World
    </div>
    <div>
      Hello World
    </div>
  </div>
  <span
    class="c1 total"
  >
    Total: $
    50
  </span>
  <button
    class="c2 checkout-btn"
  >
    Go To Checkout
  </button>
</div>
`;

exports[`should render proper cart content with mocked cart items 1`] = `
.c2 {
  min-width: 16.5rem;
  width: auto;
  height: 5rem;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 5rem;
  padding: 0 3.5rem 0 3.5rem;
  font-size: 1.5rem;
  background-color: black;
  color: white;
  text-transform: uppercase;
  font-family: 'Open Sans Condensed';
  font-weight: bolder;
  cursor: pointer;
  border: none;
}

.c2:hover {
  background-color: white;
  color: black;
  border: 1px solid black;
}

.c1 {
  text-transform: uppercase;
  font-weight: 700;
  font-size: 2.4rem;
}

.c0 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 36rem;
  height: 24rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 2rem;
  border: 1px solid black;
  background-color: white;
  top: 8rem;
  right: 0;
  z-index: 5;
}

.c0 .cart-items {
  height: 85%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: scroll;
}

.c0 .checkout-btn {
  margin-top: auto;
}

.c0 .empty-message {
  font-size: 2.2rem;
  margin: 5rem auto;
  font-weight: bold;
}

.c0 .total {
  margin-left: auto;
}

@media (min-width:768px) {
  .c0 {
    height: 34rem;
    width: 42rem;
  }
}

@media (min-width:992px) {
  .c0 {
    width: 48rem;
    height: 38rem;
  }
}

<div
  class="c0"
>
  <div
    class="cart-items"
  >
    <div>
      Hello World
    </div>
    <div>
      Hello World
    </div>
  </div>
  <span
    class="c1 total"
  >
    Total: $
    5
  </span>
  <button
    class="c2 checkout-btn"
  >
    Go To Checkout
  </button>
</div>
`;
